### PR TITLE
Add CMocka unit tests

### LINKED ISSUE #6890: Replace a few unit tests with CMocka equivalents
This task is to try to replace some of the existing DIY C unit tests in the daemon with CMocka equivalents as a means to test CMocka.

This might be a good place to start: https://github.com/netdata/netdata/blob/master/daemon/unit_test.c#L449

### PR DESCRIPTION
**Pull request by vlvkobal**
##### Summary
We are replacing some of our existing tests with ones based on the CMocka library.

The idea is to have separate tests for every individual piece of netdata in order to compile and run them fast enough when we are developing/debugging such a piece.

Closes #6890

##### Component Name
Unit tests

### CODE REVIEW

File: libnetdata/tests/test_str2ld.c [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        else if(isinf(mine))
+            assert_true(isinf(sys));
+        else if(mine != sys)
+            assert_false(abs(mine-sys) > 0.000001);
+
+        assert_ptr_equal(e_mine, e_sys);
+    }
+}
+
+int
```
Comment (thiagoftsm) [inline libnetdata/tests/test_str2ld.c]: Why is the type and function in different line here?

**Comment by vlvkobal** [comment]
To be able to run `make test` you should execute `cmake` with the `-DUNIT_TESTING=ON` option.

File: libnetdata/tests/test_str2ld.c (hunk 1 above)
Comment (vlvkobal) [reply libnetdata/tests/test_str2ld.c]: I've fixed it.

**Comment by thiagoftsm** [comment]
> To be able to run `make test` you should execute `cmake` with the `-DUNIT_TESTING=ON` option.

Thank you very much for the explanation.
After I use cmake like you said me, I could run the tests, but the output are given errors, please, see the next image:

err

**Comment by vlvkobal** [comment]
I think you forgot to run `make` before `make test`.
```
cmake -DUNIT_TESTING=ON .. && make && make test
```

**Comment by thiagoftsm** [comment]
> I think you forgot to run `make` before `make test`.
> 
> ```
> cmake -DUNIT_TESTING=ON .. && make && make test
> ```

I did not forget, the whole list of commands that I ran were:

```
  532  cd build/
  533  ls
  534  cmake -DUNIT_TESTING=ON ..
  535  make
  536  make test
  537  cd ..
  538  make test
```

When I ran the command `536` I got the error : 

```
bash-5.0$ make test
make: *** No rule to make target 'test'.  Stop.
```
so I executed `537` and when I ran again `538`  that results in  the error that I showed, I am having problem using cmocka-1.1.5 .

**Comment by vlvkobal** [comment]
What is the output of CMake?
```
...
-- Looking for CMocka which is required for unit testing
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vlad/netdata/cmake-build
```

**Comment by thiagoftsm** [comment]
bash-5.0$ cmake -DUNIT_TESTING=ON ..
-- building for: debugging
-- operating system: linux (TARGET_OS=1)
-- Judy library found
-- Checking for module 'libipmimonitoring'
--   No package 'libipmimonitoring' found
-- Checking for module 'libxenstat'
--   No package 'libxenstat' found
-- Checking for module 'snappy'
--   No package 'snappy' found
-- kinesis backend: disabled (requires AWS SDK for C++)
-- prometheus remote write backend: disabled (requires protobuf and snappy libraries)
-- mongodb backend: enabled
-- apps.plugin: enabled
-- freeipmi.plugin: disabled (depends on libipmimonitoring)
-- nfacct.plugin: enabled
-- xenstat.plugin: disabled (requires libxenstat)
-- perf.plugin: enabled
-- slabinfo.plugin: enabled
-- cgroup-network: enabled
-- Looking for CMocka which is required for unit testing
-- Configuring done
-- Generating done
-- Build files have been written to: /home/thiago/Netdata/netdata

**Comment by mfundul** [comment]
Failed to compile in Ubuntu 18.04. Did the following:
- Ran `apt-get install libcmocka-dev`
- Ran `./netdata-installer.sh`:
```
checking if LTO should be enabled... yes
checking for CMOCKA... yes
```
- Ran `make check`:
```
make  libnetdata/tests/str2ld_testdriver libnetdata/storage_number/tests/storage_number_testdriver
make[3]: Entering directory '/root/netdata'
  CC       libnetdata/tests/test_str2ld.o
  CCLD     libnetdata/tests/str2ld_testdriver
/tmp/ccM24kVp.ltrans0.ltrans.o: In function `test_str2ld':
<artificial>:(.text+0xfd): undefined reference to `_assert_true'
<artificial>:(.text+0x12f): undefined reference to `_assert_int_equal'
<artificial>:(.text+0x220): undefined reference to `_assert_true'
<artificial>:(.text+0x45f): undefined reference to `_assert_true'
/tmp/ccM24kVp.ltrans0.ltrans.o: In function `main':
<artificial>:(.text.startup+0x5b): undefined reference to `_cmocka_run_group_tests'
collect2: error: ld returned 1 exit status
Makefile:2304: recipe for target 'libnetdata/tests/str2ld_testdriver' failed
make[3]: *** [libnetdata/tests/str2ld_testdriver] Error 1
make[3]: Leaving directory '/root/netdata'
Makefile:3792: recipe for target 'check-am' failed
make[2]: *** [check-am] Error 2
make[2]: Leaving directory '/root/netdata'
Makefile:3334: recipe for target 'check-recursive' failed
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory '/root/netdata'
Makefile:3795: recipe for target 'check' failed
make: *** [check] Error 2
```

**Comment by mfundul** [comment]
Found the issue. We have gcc run this:
```
gcc -I/usr/include/uuid -g -O2 -pthread -flto -I/usr/include/uuid -lcmocka -o libnetdata/storage_number/tests/storage_number_testdriver libnetdata/storage_number/tests/test_storage_number.o libnetdata/adaptive_resortable_list/adaptive_resortable_list.o libnetdata/config/appconfig.o libnetdata/avl/avl.o libnetdata/buffer/buffer.o libnetdata/clocks/clocks.o libnetdata/dictionary/dictionary.o libnetdata/eval/eval.o libnetdata/libnetdata.o libnetdata/locks/locks.o libnetdata/log/log.o libnetdata/popen/popen.o libnetdata/procfile/procfile.o libnetdata/os.o libnetdata/simple_pattern/simple_pattern.o libnetdata/socket/socket.o libnetdata/socket/security.o libnetdata/statistical/statistical.o libnetdata/storage_number/storage_number.o libnetdata/threads/threads.o libnetdata/url/url.o libnetdata/json/json.o libnetdata/json/jsmn.o libnetdata/health/health.o -lm -lz -lcrypto -lssl -luuid -luv -llz4 -lJudy -lcrypto -lssl
```

The problem is that `-lcmocka` is before the library files and not at the end of the command.

File: libnetdata/required_dummies.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    (void)hash;
+    (void)rc;
+    (void)result;
+    return 0;
+};
+
+// required by get_system_cpus()
+char *netdata_configured_host_prefix = "";
+
+#endif // NETDATA_LIB_DUMMIES_H
```
Comment (mfundul) [inline libnetdata/required_dummies.h]: I deleted this entire file and the tests run just fine. It should be removed.

**Comment by thiagoftsm** [comment]
On Ubuntu 16 I verified that it is impossible to test, because cmake is a version not supported.

File: libnetdata/required_dummies.h (hunk 2 above)
Comment (vlvkobal) [reply libnetdata/required_dummies.h]: I can't build tests without these functions defined:
```
make  libnetdata/tests/str2ld_testdriver libnetdata/storage_number/tests/storage_number_testdriver
make[3]: Entering directory '/home/vlad/netdata'
  CC       libnetdata/tests/test_str2ld.o
  CCLD     libnetdata/tests/str2ld_testdriver
/usr/bin/ld: libnetdata/eval/eval.o: in function `eval_variable':
/home/vlad/netdata/libnetdata/eval/eval.c:170: undefined reference to `health_variable_lookup'
/usr/bin/ld: libnetdata/libnetdata.o: in function `verify_netdata_host_prefix':
/home/vlad/netdata/libnetdata/libnetdata.c:1207: undefined reference to `netdata_configured_host_prefix'
/usr/bin/ld: /home/vlad/netdata/libnetdata/libnetdata.c:1231: undefined reference to `netdata_configured_host_prefix'
/usr/bin/ld: /home/vlad/netdata/libnetdata/libnetdata.c:1235: undefined reference to `netdata_configured_host_prefix'
/usr/bin/ld: /home/vlad/netdata/libnetdata/libnetdata.c:1239: undefined reference to `netdata_configured_host_prefix'
/usr/bin/ld: /home/vlad/netdata/libnetdata/libnetdata.c:1208: undefined reference to `netdata_configured_host_prefix'
/usr/bin/ld: libnetdata/libnetdata.o:/home/vlad/netdata/libnetdata/libnetdata.c:1245: more undefined references to `netdata_configured_host_prefix' follow
/usr/bin/ld: libnetdata/log/log.o: in function `fatal_int':
/home/vlad/netdata/libnetdata/log/log.c:826: undefined reference to `send_statistics'
/usr/bin/ld: /home/vlad/netdata/libnetdata/log/log.c:828: undefined reference to `netdata_cleanup_and_exit'
/usr/bin/ld: libnetdata/os.o: in function `get_system_cpus':
/home/vlad/netdata/libnetdata/os.c:36: undefined reference to `netdata_configured_host_prefix'
/usr/bin/ld: libnetdata/os.o: in function `get_system_pid_max':
/home/vlad/netdata/libnetdata/os.c:93: undefined reference to `netdata_configured_host_prefix'
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:2443: libnetdata/tests/str2ld_testdriver] Error 1
make[3]: Leaving directory '/home/vlad/netdata'
make[2]: *** [Makefile:3939: check-am] Error 2
make[2]: Leaving directory '/home/vlad/netdata'
make[1]: *** [Makefile:3478: check-recursive] Error 1
make[1]: Leaving directory '/home/vlad/netdata'
make: *** [Makefile:3942: check] Error 2
[vlad@arch-esxi netdata]$
```

Comment (vlvkobal) [reply libnetdata/required_dummies.h]: Found it. It's the `--disable-lto` flag.
```
sudo git clean -dxf
sudo ./netdata-installer.sh --disable-lto
sudo make check
```
->  undefined references when you try building tests.

Comment (mfundul) [reply libnetdata/required_dummies.h]: I guess then we can't get rid of the file after all.

**Comment by mfundul** [comment]
> 
> 
> On Ubuntu 16 I verified that it is impossible to test, because cmake is a version not supported.

Ubuntu 18.04 LTS does not work either.

**Comment by vlvkobal** [comment]
> Ubuntu 18.04 LTS does not work either.

Look at #7091

**Comment by thiagoftsm** [comment]
I remembered this morning, on Ubuntu 16.0 I could do tests when I used autotools, the problem is with cmake.

**Pull request merged**